presto-tests/src/test/java/com/facebook/presto/resourcemanager/TestLeaderReElection.java

```diff
@@ -0,0 +1,124 @@
+        groupId = "testRaftGroupId1";
+        runner = createQueryRunner(
+                ImmutableMap.of(
+                        "resource-manager.query-expiration-timeout", "4m",
+                        "raft.isEnabled", "true",
+                        "raft.port", "6000",
+                        "raft.groupId", groupId,
+                        "raft.storageDir", "/tmp/raft-server"),
+                ImmutableMap.of(
+                        "query.client.timeout", "20s",
+                        "resource-manager.query-heartbeat-interval", "100ms",
+                        "resource-group-runtimeinfo-refresh-interval", "100ms",
+                        "concurrency-threshold-to-enable-resource-group-refresh", "0.1"),
+                ImmutableMap.of(
+                        "raft.isEnabled", "true",
+                        "raft.groupId", groupId),
+                COORDINATOR_COUNT,
+                RESOURCE_MANAGER_COUNT);
+        coordinator1 = runner.getCoordinator(0);
+        coordinator2 = runner.getCoordinator(1);
+        this.resourceManager1 = runner.getResourceManager(0);
+        this.resourceManager2 = runner.getResourceManager(1);
+        runner.getCoordinators().stream().forEach(coordinator -> {
+            coordinator.getResourceGroupManager().get().addConfigurationManagerFactory(new FileResourceGroupConfigurationManagerFactory());
+            coordinator.getResourceGroupManager().get()
+                    .setConfigurationManager("file", ImmutableMap.of("resource-groups.config-file", getResourceFilePath(RESOURCE_GROUPS_CONFIG_FILE)));
+        });
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void teardown()
+    {
+        closeQuietly(coordinator1);
+        closeQuietly(coordinator2);
+        closeQuietly(resourceManager1);
+        closeQuietly(resourceManager2);
+        closeQuietly(client);
+        coordinator1 = null;
+        coordinator2 = null;
+        resourceManager1 = null;
+        resourceManager2 = null;
+        client = null;
+    }
+
+    private String getResourceFilePath(String fileName)
+    {
+        return this.getClass().getClassLoader().getResource(fileName).getPath();
+    }
+
+    @Test(timeOut = 120_000)
+    public void testRaftServerForResourceManager()
+            throws Exception
+    {
+        RatisClient client = coordinator1.getRatisClient();
+        LeaderResourceManagerAddressSelector selector = new LeaderResourceManagerAddressSelector(resourceManager1.getNodeManager(), client);
+        Optional<SimpleAddressSelector.SimpleAddress> addressBeforeShutdown = selector.selectAddress(Optional.empty());
+        assertTrue(addressBeforeShutdown.isPresent());
+
+        if (client.getLeader().equals(resourceManager1.getNodeManager().getCurrentNode().getNodeIdentifier())) {
+            resourceManager1.close();
+        }
+        else {
+            resourceManager2.close();
+        }
+
+        client.sendMessage();
+        sleep(10000);
```

Comment:
do we really need this sleep anymore? If yes, can we reduce that to a lower value if possible?

Reply:
I agree. Is there a listener I can register to get notified of Leader Elections? . Does Ratis provide any such capabilities.?

Reply:
There isn't a listener but if the client is in communication with the leader it becomes aware of the new leader once the leader election is complete.